libs/core/type_support/include/hpx/type_support/uninitialized_relocate.hpp

```diff
@@ -0,0 +1,175 @@
+                choose_uninitialized_relocate_helper<InIter, FwdIter>::value ==
+                    relocate_strategy::buffer_memcpy,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst) noexcept
+        {
+            auto n_objects = std::distance(first, last);
+
+            if (n_objects != 0)
+            {
+                void* first_void = const_cast<void*>(
+                    static_cast<const void*>(std::addressof(*first)));
+                void* last_void = const_cast<void*>(
+                    static_cast<const void*>(std::addressof(*last)));
+
+                void* dst_void = const_cast<void*>(
+                    static_cast<const void*>(std::addressof(*dst)));
+
+                std::byte* first_byte =
+                    reinterpret_cast<std::byte*>(first_void);
+                std::byte* last_byte = reinterpret_cast<std::byte*>(last_void);
+
+                auto n_bytes = std::distance(first_byte, last_byte);
+
+                // Ideally we would want to convey to the compiler
+                // That the new buffer actually contains objects
+                // within their lifetime. But this is not possible
+                // with current language features.
+                std::memmove(dst_void, first_void, n_bytes);
+
+                dst += n_objects;
+            }
+
+            return dst;
+        }
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<
+                choose_uninitialized_relocate_helper<InIter, FwdIter>::value ==
+                    relocate_strategy::for_loop_nothrow,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst) noexcept
+        {
+            for (; first != last; ++first, ++dst)
+            {
+                // the move + destroy version will be used
+                hpx::experimental::relocate_at(
+                    std::addressof(*first), std::addressof(*dst));
+            }
+
+            return dst;
+        }
+
+        template <typename InIter, typename FwdIter,
+            std::enable_if_t<
+                choose_uninitialized_relocate_helper<InIter, FwdIter>::value ==
+                    relocate_strategy::for_loop_try_catch,
+                int> = 0>
+        FwdIter uninitialized_relocate_helper(
+            InIter first, InIter last, FwdIter dst)
+        {
+            FwdIter original_dst = dst;
+
+            for (; first != last; ++first, ++dst)
+            {
+                try
+                {
+                    // the move + destroy version will be used
+                    hpx::experimental::relocate_at(
+                        std::addressof(*first), std::addressof(*dst));
+                }
+                catch (...)
+                {
+                    // destroy all objects other that the one
+                    // that caused the exception
```

Comment:
`s/that/than/`
The one that caused the exception was already destroyed, by `relocate_at`. So I might say
```
 // destroy all objects other than the one that caused
 // the exception (relocate_at destroyed that one)
```

Reply:
Ohhh I had not understood that exceptions are the reason for using destroy guards instead of just move-constructing and destroying directly. Thanks, I will change the comment to reflect that